pycozmo/audio.py

```diff
@@ -0,0 +1,147 @@
+import struct
+import time
+import wave
+from datetime import datetime, timedelta
+from threading import Thread, Lock
+
+from .conn import ClientConnection
+from .protocol_encoder import OutputAudio, Keyframe
+
+MIN_WAIT = 0.033
+
+class AudioManager():
+    """
+    This class takes care of reading audio files and generating the OutputAudio messages sent to
+    cozmo.
+
+    The play() method can be used to play an audio file or list of OutputAudio messages.
+
+    Args:
+        conn (ClientConnection): client managing the communication with the robot
+    """
+    def __init__(self, conn: ClientConnection):
+        self.stream = []
+        self.stop = False
+        self.conn = conn
+        self.thread = Thread()
+        self.lock = Lock()
+        self.audio_gen = []
+        self.conn.add_handler(Keyframe, self._on_keyframe)
+
+    def start(self):
+        self.stop = False
+        if not self.thread.is_alive():
+            self.thread = Thread(target=self.run, name=__class__.__name__)
+            self.thread.start()
+
+    def stop(self) -> None:
+        self.stop = True
+        self.thread.join()
+        self.audio_gen = []
+
+    def run(self) -> None:
+        while len(self.audio_gen) > 0:
+            generator = []
+
+            with self.lock:
+                generator = self.audio_gen.pop(0)
+
+            next_trigger_time = datetime.now()
+
+            for pkt in generator:
+
+                resting_time = (next_trigger_time - datetime.now()).total_seconds()
+                if resting_time > 0:
+                    time.sleep(resting_time)
+
+                self.conn.send(pkt, True)
+
+                next_trigger_time = datetime.now() + timedelta(seconds=MIN_WAIT)
+
+                if self.stop:
+                    return
+
+
+    def _play_wav(self, filename: str):
+        with wave.open(filename, "r") as w:
+            sampwidth = w.getsampwidth()
+            ratediv = 2 if w.getframerate() > 30000 else 1
+            channels = w.getnchannels()
+            done = False
+            while not done:
+                frame = bytes_to_cozmo(w.readframes(744 * ratediv),
+                                            sampwidth, ratediv, channels)
+
+                if len(frame) < 744:
+                    frame += [0] * (744 - len(frame))
+                    done = True
+
+                yield OutputAudio(frame)
+
+    def _play_packets(self, packets: list):
+        for p in packets:
+            yield p
+
+    def play(self, audio):
+        if '.wav' in audio:
```

Comment:
Hmmm you are right, I thought this wouldn't be a problem since this comparison can be done as long as `audio` is a list, tuple or string, but that would prevent having a list of strings that would give an error. Good point, thanks!